server/schedulers/hot_region.go

```diff
@@ -490,29 +512,61 @@ func (bs *balanceSolver) allowBalance() bool {
 	}
 }
 
-func (bs *balanceSolver) selectSrcStoreID() uint64 {
-	var id uint64
-	switch bs.opTy {
-	case movePeer:
-		id = selectSrcStoreByByteRate(bs.stLoadDetail)
-	case transferLeader:
-		if bs.rwTy == write {
-			id = selectSrcStoreByCount(bs.stLoadDetail)
-		} else {
-			id = selectSrcStoreByByteRate(bs.stLoadDetail)
+func (bs *balanceSolver) filterSrcStores() map[uint64]*storeLoadDetail {
+	ret := make(map[uint64]*storeLoadDetail)
+	for id, detail := range bs.stLoadDetail {
+		if bs.cluster.GetStore(id) == nil {
+			log.Error("failed to get the source store", zap.Uint64("store-id", id))
+			continue
 		}
+		if len(detail.HotPeers) == 0 {
+			continue
+		}
+		ret[id] = detail
 	}
-	if id != 0 && bs.cluster.GetStore(id) == nil {
-		log.Error("failed to get the source store", zap.Uint64("store-id", id))
-	}
-	return id
+	return ret
 }
 
-func (bs *balanceSolver) getPeerList() []*statistics.HotPeerStat {
-	ret := bs.stLoadDetail[bs.srcStoreID].HotPeers
-	bs.sche.r.Shuffle(len(ret), func(i, j int) {
-		ret[i], ret[j] = ret[j], ret[i]
+func (bs *balanceSolver) getHotPeers() []*statistics.HotPeerStat {
+	ret := bs.stLoadDetail[bs.cur.srcStoreID].HotPeers
+	if len(ret) <= maxPeerNum {
+		return ret
+	}
+
+	byteSort := make([]*statistics.HotPeerStat, len(ret))
+	copy(byteSort, ret)
+	sort.Slice(byteSort, func(i, j int) bool {
+		return byteSort[i].GetByteRate() > byteSort[j].GetByteRate()
+	})
+	keySort := make([]*statistics.HotPeerStat, len(ret))
+	copy(keySort, ret)
+	sort.Slice(keySort, func(i, j int) bool {
+		return keySort[i].GetKeyRate() > keySort[j].GetKeyRate()
 	})
+
+	union := make(map[*statistics.HotPeerStat]struct{}, maxPeerNum)
+	for len(union) < maxPeerNum {
```

Comment:
why need the for loop?

Reply:
To prevent return too many peers, which will make `balanceSolver.solve()` too slow.

Reply:
Got.